elytron/WFCORE-5178_native-support-for-oidc.adoc

```diff
@@ -0,0 +1,164 @@
+should be used instead.
+
+The new HTTP mechanism implementation will refresh the access token when it is expired. Since the Keycloak client adapter
+provides an option to refresh the token on every request, we should provide this option as well but make
+sure to document that the use of this option is discouraged since it can have a significant impact on performance.
+
+==== New elytron-oidc subsystem
+
+A new subsystem, `elytron-oidc` will be added to WildFly. Its schema will be similar to the existing
+https://github.com/keycloak/keycloak/blob/master/adapters/oidc/wildfly/wildfly-subsystem/src/main/resources/schema/wildfly-keycloak_1_1.xsd[keycloak-adapter-subsystem].
+The new subsystem will install a `DeploymentUnitProcessor` which will detect if OIDC is required for a deployment.
+
+Like with the existing Keycloak client adapter, the OIDC configuration will be specified via the deployment
+or via subsystem configuration. For the deployment case, the OIDC configuration will be specified in a JSON
+configuration file in the WEB-INF directory along with the `auth-method`, `OIDC`, specified in the `web.xml`.
+For the subsystem case, the `auth-method` won't need to be specified in the `web.xml` file, it will be set
+automatically.
+
+Once a deployment has been processed, if its `auth-method` is `OIDC`, the OIDC integration will be activated.
+Otherwise, no activation will occur and the deployment will continue to use the existing security capabilities
+provided by WildFly.
+
+If the OIDC integration is activated for a deployment, then appropriate module dependencies will be added to
+the deployment (e.g., a dependency on the new Elytron OIDC module). This will ensure the activation of the
+OIDC HTTP mechanism implementation for the deployment. The deployment will make use of a "Virtual Security Domain",
+similar to what was implemented for https://github.com/wildfly/wildfly-proposals/blob/master/microprofile/WFLY-12609_MicroProfile_JWT_1_1.adoc#hard-requirements[Microprofile JWT].
+This means that it will no longer be necessary to pre-configure SecurityDomains in the Elytron subsystem,
+as was required with the Keycloak client adapter.
+
+It should be possible to make use of different OpenID providers with WildFly. While working on the
+implementation for this RFE, we should make sure that we aren't including Keycloak-specific configuration, if
+possible.
+
+==== New galleon layer for OIDC support
+
+A new Galleon layer, `oidc`, should be added to allow a server to be provisioned with OIDC support.
+
+=== Nice-to-Have Requirements
+
+https://openid.net/specs/openid-connect-rpinitiated-1_0.html[RP-Initiated Logout] is an optional
+OIDC specification that defines how a relying party (RP) can request that an OpenID provider logs
+out the end-user. The Keycloak client adapter provides support for this so this is something we
+should provide as well but we could possibly look at that as a follow-up to this RFE if necessary.
+
+=== Non-Requirements
```

Comment:
Do we want to deal with migration paths from the Keycloak client adapter to the new subsystem?

Reply:
The same subsystem attributes and JSON configuration options that could be used with the `keycloak-adapter-subsystem` and the `keycloak.json` file will still work with the `elytron-oidc-client` subsystem and the `oidc.json` file. The only difference is the actual subsystem name and the name of the JSON configuration file and auth-method. 

That means that the migration tool could potentially detect an existing `keycloak` subsystem configuration and then change its name to the new subsystem name instead. I can add this as a nice to have.

For the deployment side, do you think it would make sense for the new subsystem to support both the old JSON configuration file name and `auth-method` (i.e., keycloak.json and `<auth-method>KEYCLOAK</auth-method>` in addition to the new JSON configuration file name and `auth-method` (i.e., `oidc.json` and `<auth-method>OIDC</auth-method>`). That way, a user wouldn't need to update their deployment configuration to switch to the new approach?

Reply:
I don't know how much do we want to push this feature.

In case it's just an alternative to the Keycloak client adapter and the adapter support will be kept, the migration step in the tool should be rather an optional step (to be explicitly enabled). As the migration tool enables to migrate also non-standard modules/layers, it should be also possible do disable the migration of any Keycloak client adapter related modules.

In case we want to push this feature, or we know that the WF Keycloak client adapter will be deprecated soon (e.g. because of this feature), we should provide the migration support and enable it by default.

I agree that it would be great to support Keycloak-secured deployments without a change. However supporting Keycloak configuration and `auth-method` in deployments by the new subsystem seems a bit dirty and could bring some complexity to handle cases when having both the subsystems in use (`keycloak-adapter-subsystem` and `elytron-oidc-client`). Maybe it could be possible and sufficient if the migration tool changed the the `auth-method` and the JSON configuration file name during a deployment migration.

Reply:
Thanks, @OndrejKotek.

The idea is that we won't need the Keycloak client adapter anymore, only this feature will be used.

Having the migration tool update the `auth-method` and JSON configuration file name during a deployment migration is a good idea.

I'll update the document to reflect this.

Reply:
@fjuma Will we allow the two subsystems to work in parallel?

@emmartins Is it possible to adjust the migration tool to update `auth-method` in `web.xml` and rename a file during a deployment migration? I suppose we haven't performed a similar step/operations so far, but I also don't see an obstacle to have such thing.

Reply:
> @fjuma Will we allow the two subsystems to work in parallel?

Nothing would prevent them from working in parallel. But the idea will be to switch fully to the new subsystem so the Keycloak adapter won't need to be produced anymore at some point.

Reply:
> The idea is that we won't need the Keycloak client adapter anymore, only this feature will be used.

Are Keycloak developers aware of this idea, so that they could drop the support on their side? A KEYCLOAK Jira issue?

Reply:
They are aware that we are working on this. We'll need to first deprecate the existing adapter and then remove later on.

Reply:
@OndrejKotek @fjuma The migration tool does not touches deployments, it just provides the feature to copy deployments to the new server. Any app/deployment migration issue should be handled by Migration Toolkit for Applications (MTA) and our Migration Guide.

Reply:
Wrt the subsystem (name) migration, agree that if supported by the tool such migration task it should be off by default.